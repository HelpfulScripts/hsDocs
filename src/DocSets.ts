/**
 * DocSets.ts. Loads the doc.json files to process and display as documentation.
 * Processing occurs in these steps:
 * 1. Load the index.json file that contains a list of doc.json files to load, one for each library to show
 * 2. Load each doc.json file, which describes a library
 * 3. Call DocSets.add to add the library name to the registry and create an index of entries for the library
 */

/** */
import { m }            from 'hslayout';
import { Log }          from 'hsutil'; const log = new Log('DocsSets');
import { DocsNode }     from './Nodes';

const DOCDIR:string = './data/';

/**
 * the default location for the index .json files, relative to the web page:
 * `'./data/index.json'`
 */
const FILE:string = './data/index.json';

export type json = any;

function matchAll(str:string, re:RegExp): string[] {
    const result:string[] = [];
    let partial:string[];
    while ((partial = re.exec(str)) !== null) {
        result.push(partial[1]);
        log.debug(log.inspect(partial));
    }
    return result;
}

/**
 * DocSets object. Keeps a list of registered docsets and 
 * provides access to elements of each docset.
 */
export class DocSets { 
    /** Contains references to the docsets and all elements per docset, accessible per ID. */
    // private static gList:List = {index:{}};

    private static libs:string[] = [];
    private static docs:string[] = [];

    /** an optional title for the set of DocSets loaded, as specified in the `index.json` file. */
    private static gTitle: string = '';
    private static nodeList = <{[id:string]: DocsNode}>{};
    
    public static nodeCount = 0;
    
    public static addNode(mdl:json, node:DocsNode) {
        const id = mdl.id;
        const lib =  mdl.lib;
        DocSets.nodeList[`${lib}.${id}`] = node;
        DocSets.nodeList[mdl.fullPath] = node;
// if (mdl.fullPath.indexOf('hsDocs.DocSets')>=0) { log.info(`added node ${id}: '${mdl.fullPath}'`); }
        DocSets.nodeCount++;
    }


    /**
     * Adds the docset in `content` to the `gList` at the position of `file` in `DocSets.docs`.
     * @param content the docSet content to add. 
     * @param file the file name from which the content was read. 
     */
    public static addDocSet(content:json, file:string):void {
        const lib = content.name;
        const i = DocSets.docs.indexOf(file);
        DocSets.libs[i] = lib;
        const root = DocsNode.traverse(content, lib);
        log.info(`traversed '${root.fullPath}'`);
    }
    
    public static getNode(id:string|number, lib:string):DocsNode {
        const key = ((typeof id === 'number') || (parseInt(''+id, 10)>=0))? `${lib}.${id}` : id;
        // log.info(`getNode id=${id}, lib=${lib} -> ${key}, typeof=${typeof id}, parseInt=${parseInt(''+id, 10)}`);
        if (DocSets.nodeList.length && !DocSets.nodeList[key]) { 
            log.warn(`did not find node for key '${key}' (id=${id}, lib=${lib})`); 
            log.warn(new Error().stack);
        }
        return DocSets.nodeList[key]; 
    }
    
    /**
     * returns the specified documentation element.
     * When called without parameters, a `string[lib]` of available docSets is returned.
     * When called with only `lib` specified, the corresponding docSet overview is returned.
     * @param lib specifies the docset to scan. 
     * @param id specifies the element within the docSet, either by its id number, or by its fully qualified path, 
     * e.g. 'hsDocs.DocSets.DocSets.add' 
     */
    public static getLibs():string[]{ 
        return DocSets.libs; 
    }

    /**
     * loads an index set and the docsets it contains from directory `dir`.
     * @param file the optional directory to read from. If unspecified, 
     * read from the index file specified by {@link DocSets.FILE `FILE`}.
     * @return a promise to load the index set
     */
    public static async loadList(file?:string):Promise<void> {
        /** load an `index.json` file that contains references to the DocSets to load. */
        async function getIndexFile(dir:string, url:string):Promise<boolean> {
            if (!url) { return false; }
            try {
                const result = await m.request({ method: "GET", url: file});
                DocSets.gTitle = result.title;
                DocSets.docs = result.docs.map((doc:string) => doc.indexOf(':')>0? doc : dir+doc);
                log.info(`found index file ${url} with ${DocSets.docs.length} library references`);
                return true;
            } catch(e) {
                return false;
            }
        }
        /** 
         * get all `jsons` in a directory specified by `url`. 
         * If the directory contains an `index.json`, calls `getIndexFile` on that file's url.
         */
        async function getDirJSONs(url:string):Promise<boolean> {
            let result = false;
            await m.request({ method: "GET", url: url, 
                // catch extract to avoid attempt to deserialize
                extract: async (xhr:any, options:any) => { 
                    const jsons = matchAll(xhr.responseText, /<a href="(\S*?.json)">/g);  // all *.json references
                    // check if index.jon exists:
                    const index = jsons.filter(m => m==='index.json')[0];
                    // const jsons = matches.filter(m => !m.startsWith('index'));

                    if (index) { 
                        result = await getIndexFile(dir, index); 
                    } else if (jsons.length) {
                        log.info(`found dir list in ${url}`);
                        DocSets.gTitle = '';
                        DocSets.docs = jsons;
                        result = true;
                    }
                }
            });
            return result;
        }
        let i = file.lastIndexOf('/');
        const dir = file.substring(0,i+1);
        
        let found = false;
        if (!found) { found = await getIndexFile(dir, file); }
        if (!found) { found = await getDirJSONs(dir); }
        if (!found) { found = await getDirJSONs(DOCDIR); }
        log.debug(`found ${DocSets.docs.length} dos sets: ${log.inspect(DocSets.docs, 5)}`);
        await Promise.all(DocSets.docs.map(async (f:string) => await loadDocSet(f))).catch(log.error);        
        log.info(`found ${DocSets.nodeCount} DocNodes`);
        m.redraw();
    }
}

/**
 * Loads a docset specified by file from the directory `dir`. 
 * Once received, the docset is registered in `modules` via the `add` method.
 * @param dir the directory to read from
 * @param file the `json` file to load as docset
 */
async function loadDocSet(file:string):Promise<void> {
    log.debug(`loading ${file}`);
    const r:json = await m.request({ method: "GET", url: file });
    DocSets.addDocSet(r, file);
}

