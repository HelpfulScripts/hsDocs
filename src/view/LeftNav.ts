/**
 * LeftNav: Responsible for constructing the left-hand navigation pane. 
 */

/** */
import m from "mithril";
import { Vnode }        from 'hswidget';
import { Widget }       from 'hswidget';
import { Collapsible }  from 'hswidget';
import { Log }          from 'hsutil';  const log = new Log('LeftNav');
import { libLinkByPath }from '../NodesDisplay'; 
import { DocSets }      from '../DocSets';
import { DocsNode  }    from "../Nodes";
import { ModuleDesc }   from "../Nodes";
import { Group }        from "../Nodes";
import { DocsAttrs }    from "./DocsMenu";
 

interface LeftNavAttrs extends DocsAttrs {
}

/**
 * Constructs the left-hand navigation pane
 */
export class LeftNav extends Widget {
    view(node:Vnode<LeftNavAttrs, this>) {
        const lib = node.attrs.lib;
        const field = node.attrs.field;
        const docSet = DocSets.getNode(0, lib) || {name:'unknown', id:0};
        return m('.hs_left', node.attrs, navList(<DocsNode>docSet, field));
    }
}

/** creates the list if modules (`*.ts` files) */
function navList(docSet:DocsNode, field:string):m.Children {
    /** processes a module, i.e. a `.ts` file. */
    function externalModule(mdl:ModuleDesc) {
        let selected = false;
        // if (field===''+mdl.id || field.indexOf(mdl.fullPath) === 0) { selected=true; }
        if (field.indexOf(mdl.fullPath) === 0) { selected=true; }

        return m(Collapsible, {class:`hs_left_nav_module`, preArrow: true, isExpanded:selected}, [
            m(`${selected?'.hs_left_nav-selected':''}`, libLinkByPath(mdl.lib, mdl.fullPath, mdl.name, `.hs_left_nav_module_name`)),
            ...(mdl.groups || []).map((g:any) => entries(g, mdl, field))
        ]);
    }

    if (docSet.kind === 0) { // External DocSets
        collectModules(docSet);
        const modules = docSet.modules.map(externalModule);
        modules.unshift(m('.hs_left_nav-header', 'Modules'));
        return m('.hs_left_nav', [m('.hs_left_nav-content', modules)]);
    }
}

function collectModules(docSet:DocsNode) {
    const modulesByName = <ModuleDesc>{};
    docSet.modules = <ModuleDesc[]>[];
    if (docSet.children) {
        docSet.children.forEach(c => {
            if (!c) { 
                log.warn(`c is undefined for docSet ${docSet.id}`);
                return;
            }
            // don't show modules from other projects (isExternal flag) or modules on the ignore list
            if (!(c.flags && c.flags.isExternal) && !ignoreModules[c.name]) {
                // const name = c.innerModule? c.innerModule : c.name;
                const name = c.name;
                let module = modulesByName[name];
                if (!module) {  // new module
                    docSet.modules.push(module = modulesByName[name] = <ModuleDesc>{ 
                        name: name,
                        lib: docSet.lib,
                        fullPath: docSet.fullPath + '.'+ name,
                        groups:[]
                    });
                }
                // get existing module groups
                const groups = <{[title:string]:Group}>{};
                module.groups.forEach((g:Group) => groups[g.title] = g); 
                // for each group in child:
                if (c.groups) { c.groups.forEach((g:any) => {
                    let group = groups[g.title]; // get existing 
                    if (!group) {                 //  else create new
                        group = groups[g.title] = <Group>{
                            children:[],
                            kind: g.kind,
                            title: g.title
                        };
                    module.groups.push(group);
                    }
                    group.children = group.children.concat(g.children);
                });}
            }
        });
    }
}


/**
 * modules to ignore in the list
 */
const ignoreModules = {
    overview:   true,   // the project overview.ts file
    index:      true    // the index.ts file
};

//const expanded: {string?:boolean} = {};

/**
 * processes a group of entries, e.g. Variables, Functions, or Classes.
 * @param group the group structure within the docset
 * @param mdl the entire module docset
 * @param field the field ID to be displayed on the main panel
 */
function entries(group:any, mdl:any, field:string) {
    function moduleGet(c:any) {
        return DocSets.getNode(c, mdl.lib);
    }
    /**
     * processes one entry within a group, e.g. one variable, function, or class.
     */
    function entry(mod:any) { 
        const selected = (field===''+mod.id || field===mod.fullPath)? '.hs_left_nav-selected' : '';
        const exported = (mod.flags && mod.flags.isExported);
        const isStatic = (mod.flags && mod.flags.isStatic);
        const css = `.hs_left_nav_entry_link ${selected} ${exported?'.hs_left_nav-exported' : ''}`;
        return (!exported && group.title==='Variables')? '' :   // ignore local module variables
            m('.hs_left_nav_entry', [
                isStatic? 'static': '',
                libLinkByPath(mod.lib, mod.fullPath, mod.name, css)
            ]);
    }

    function empty(mod:any) { return mod !== ''; }

    let grp = [];
    if (group && group.children) {
        grp = group.children
            .map(moduleGet)         // replace id reference by module
            .sort(exportAscending)  // sort: exported first, then alphabetically
            .map(entry)             // replace module by vnode structure
            .filter(empty);         // filter empty elements
        if (grp.length > 0) { // add an entries header if there are elements
            grp.unshift(m('.hs_left_nav-header', group.title));
        }
    }
    return (grp.length > 1)? m(`.hs_left_nav_entries`, grp) : '';
}

/**
 * sorting function: sort first by exported status, then alphabetically.
 */
function exportAscending(a:any, b:any):boolean|number {
    if (a.flags && b.flags) {
        if (a.flags.isExported && b.flags.isExported) { return a.name > b.name; }
        else if (a.flags.isExported) { return -1; }
        else if (b.flags.isExported) { return 1; }
        else { return a.name > b.name; }
    } else { return a.name > b.name; }
}

