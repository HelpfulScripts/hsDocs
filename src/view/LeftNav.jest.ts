import { LeftNav } from './LeftNav';
import { m } from 'hslayout';
import { DocSets } from '../DocSets';
const fs = require('fs');


const set = {
    "docs": [
        "hsDocs.json"
    ],
    "title": "HS Libraries"
};

m.request = (req: any) => {
    if (req.url === 'mylist.json') {
        return Promise.resolve(set);
    } else if(req.url.endsWith('.json')) {
        return new Promise((resolve:(data:any)=>void, reject:(err:any)=>void) => {
            fs.readFile(`${__dirname}/../example/${req.url}`, 'utf8', (err:any, data:any) => {
                if (err) { throw err; }
                else { resolve(data); }
            });
        })
        .then((data:any) => (typeof data === 'string')? JSON.parse(data) : data)
        .catch(console.log)
        ;
    } else {
        return Promise.resolve('dir');
    }
};

m.route.param = (name:string) => {
    if (name==='field') { return '0'; }
    if (name==='lib')   { return 'hsDocs'; }
    return '';
};

const root:any = window.document.createElement("body");


describe('LeftNav', () => {
    beforeAll((done) => { 
        DocSets.loadList('mylist.json')
        .then(() => {
            m.mount(root, { view: () => { return m(LeftNav, { route:{lib: 'hsDocs', field:'0'}});} }); 
            setTimeout(() => {
                m.redraw();
                done();
            }, 500);
        });
    });

    it('should exist', (done)=>{
        expect(LeftNav).toBeDefined();
        done();
    });

    it('shoud match snapshot', (done) => {
        expect(root).toMatchSnapshot();
        done();
    });
});


