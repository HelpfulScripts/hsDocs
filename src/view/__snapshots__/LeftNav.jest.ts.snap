// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftNav overview matches snapshot 1`] = `
<body>
  <div
    class="hs-layout  "
  >
    <div
      class="hs-left"
    >
      <div
        class="hs-left-nav"
      >
        <div
          class="hs-left-nav-content"
        >
          <div
            class="hs-left-nav-header"
          >
            Modules
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.DocSets"
                >
                  DocSets
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Classes
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.DocSets.DocSets"
                    >
                      DocSets
                    </a>
                  </div>
                </div>
              </div>
              <div>
                
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.DocSets.loadDocSet"
                    >
                      loadDocSet
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.DocSets.recursiveIndex"
                    >
                      recursiveIndex
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.Site"
                >
                  Site
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.Site.init"
                    >
                      init
                    </a>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Object literals
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.Site.myConfig"
                    >
                      myConfig
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.markdown"
                >
                  markdown
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.markdown.markDown"
                    >
                      markDown
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.markdown.substituteLinks"
                    >
                      substituteLinks
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.DocsMenu"
                >
                  view.DocsMenu
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Classes
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.DocsMenu.DocsMenu"
                    >
                      DocsMenu
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.LeftNav"
                >
                  view.LeftNav
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Classes
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.LeftNav.LeftNav"
                    >
                      LeftNav
                    </a>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.LeftNav.entries"
                    >
                      entries
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.LeftNav.exportAscending"
                    >
                      exportAscending
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.LeftNav.navList"
                    >
                      navList
                    </a>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Object literals
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.LeftNav.ignoreModules"
                    >
                      ignoreModules
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.MainComment"
                >
                  view.MainComment
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.MainComment.comment"
                    >
                      comment
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.MainComment.commentLong"
                    >
                      commentLong
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainComment.commentRemainder"
                    >
                      commentRemainder
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainComment.mainCommentParams"
                    >
                      mainCommentParams
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainComment.otherCommentTags"
                    >
                      otherCommentTags
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainComment.prettifyCode"
                    >
                      prettifyCode
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainComment.returns"
                    >
                      returns
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainComment.textOrShortTextOrDescription"
                    >
                      textOrShortTextOrDescription
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.MainDetail"
                >
                  view.MainDetail
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Classes
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.MainDetail"
                    >
                      MainDetail
                    </a>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.getOverview"
                    >
                      getOverview
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.itemChild"
                    >
                      itemChild
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.itemDescriptor"
                    >
                      itemDescriptor
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.itemDoc"
                    >
                      itemDoc
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.member"
                    >
                      member
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.members"
                    >
                      members
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.overviewDoc"
                    >
                      overviewDoc
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.parameter"
                    >
                      parameter
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainDetail.title"
                    >
                      title
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.MainExample"
                >
                  view.MainExample
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Interfaces
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.CommentDescriptor"
                    >
                      CommentDescriptor
                    </a>
                  </div>
                </div>
              </div>
              <div>
                
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.MainExample.example"
                    >
                      example
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.addExample"
                    >
                      addExample
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.addExampleStructure"
                    >
                      addExampleStructure
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.createExecuteScript"
                    >
                      createExecuteScript
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.executeError"
                    >
                      executeError
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.executeScript"
                    >
                      executeScript
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.getCommentDescriptor"
                    >
                      getCommentDescriptor
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.getNewID"
                    >
                      getNewID
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.initDesc"
                    >
                      initDesc
                    </a>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Object literals
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry  "
                      href="#!/api/hsDocs/hsDocs.view.MainExample.modules"
                    >
                      modules
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.Parts"
                >
                  view.Parts
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                
              </div>
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.kindString"
                    >
                      kindString
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.defaultVal"
                    >
                      defaultVal
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.extensionOf"
                    >
                      extensionOf
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.flags"
                    >
                      flags
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.inheritedFrom"
                    >
                      inheritedFrom
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.itemName"
                    >
                      itemName
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.itemTooltip"
                    >
                      itemTooltip
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.libLink"
                    >
                      libLink
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.makeID"
                    >
                      makeID
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.signature"
                    >
                      signature
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.sourceLink"
                    >
                      sourceLink
                    </a>
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Parts.type"
                    >
                      type
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="hs-collapsible  hs-left-nav-module"
          >
            <div
              class="hs-collapsible-title"
            >
              <div
                class="hs-collapsible-pre  hs-collapsible-arrow-right"
              />
              <div>
                <a
                  class="hs-left-nav-module-name "
                  href="#!/api/hsDocs/hsDocs.view.Tooltip"
                >
                  view.Tooltip
                </a>
              </div>
              <div />
            </div>
            <div
              class="hs-collapsible-content "
            >
              <div>
                <div
                  class="hs-left-nav-entries"
                >
                  <div
                    class="hs-left-nav-header"
                  >
                    Functions
                  </div>
                  <div>
                    
                    <a
                      class="hs-left-nav-entry   hs-left-nav-exported"
                      href="#!/api/hsDocs/hsDocs.view.Tooltip.tooltip"
                    >
                      tooltip
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
